### PR TITLE
[Enhancement] Add an opt-out build flag for the Lance connector

### LINKED ISSUE #74665: [Feature] Native Lance connector for StarRocks
- [x] https://github.com/StarRocks/starrocks/pull/74659
## Is your feature request related to a problem? Please describe.

StarRocks currently supports external table connectors for Iceberg, Delta Lake, Hudi, Paimon, and other formats. However, there is no support for [Lance](https://github.com/lancedb/lance), a modern columnar data format optimized for ML workloads — embedding vectors, multi-modal data, and large-scale training datasets.

Users who store datasets in Lance format must currently ETL data into a supported format before querying with StarRocks, adding latency and operational complexity.

## Describe the solution you'd like

A native Lance connector that allows StarRocks to directly query Lance datasets on local and cloud filesystems (S3, ADLS Gen2, GCS), enabling:

- Ad-hoc exploration of ML training datasets (embeddings, document corpora)
- Quality analysis by joining Lance datasets with metadata in Iceberg/Hive tables
- Interactive analytics replacing Spark read paths on Lance tables

## Implementation Plan

The implementation is split into independently reviewable PRs:

| PR | Scope | Status |
|----|-------|--------|
| #74659 | Thrift definitions for Lance scan range fields | Open |
| #74660 | FE optimizer operators and rules | Open |
| #74661 | FE scan node and plan fragment builder | Open |
| #74662 | Java-extensions Lance JNI reader | Open |
| #74663 | BE optional pluggable connector module (follows benchmark pattern) | Open |
| #74664 | Build flag `--without-connector-lance` | Open |

**Already merged (Phase 1):** FE catalog & metadata discovery (`LanceConnector`, `LanceMetadata`, `LanceApiConverter`, `LanceTable`)

## Architecture

- **FE**: `LANCE` registered as `ConnectorType` and `TableType`. Query planning reuses `THdfsScanRange` with Lance-specific fields. CBO statistics integration.
- **BE**: Optional pluggable module following the benchmark connector pattern (`--without-connector-lance` compile-time opt-out). JNI scanner delegates to `lance-core` Java SDK.
- **java-extensions**: `LanceSplitScanner` using Arrow vectors via `lance-core:7.0.0`

## Future Work

- Predicate pushdown expression visitor
- Cloud storage credential passthrough (Azure Workload Identity, S3 IAM, GCS)
- HMS-backed catalog discovery (replacing property-based bootstrap)
- Native Rust FFI scanner (zero-copy via Arrow C Data Interface)

## Are you willing to submit PR?

Yes — PRs are already submitted (see table above).

### PR DESCRIPTION
**Pull request by beinan**
## Why I'm doing:

Builds that opt out of Lance must omit the native module, Maven reader module, and packaged reader artifacts consistently.

## What I'm doing:

Support Arrow LargeUtf8/LargeBinary and widen unsigned integers without overflow (UInt8/16/32/64 to SMALLINT/INT/BIGINT/DECIMAL(20,0)); align FE schema declarations with reader decoding.

Add `--without-connector-lance` and propagate it to CMake, Maven profile selection, and BE packaging. The default build includes Lance. An enabled build fails if reader artifacts are missing instead of silently producing an incomplete package.

Packaging is introduced alongside the reader in #74662. Column pruning and date64 fixes requested in review are placed in the optimizer and reader layers respectively and are included in this cumulative branch.

Series order: #74660 → #74661 → #74662 → #74663 → #74664. Catalog and Thrift prerequisites (#74420 and #74659) are already merged. These branches are cumulative against `main`; each preceding layer is required.

### Validation

- CMake configuration succeeds with Lance ON and with Lance OFF plus dynamic modules; the OFF compilation database contains no Lance connector source.
- Maven effective POM includes the reader by default and excludes it with `skipLanceReader`.
- Package checks verify both ON JAR destinations, OFF exclusion, exclusion of test JARs, and failure for missing enabled artifacts; `bash -n build.sh` passes.
- Completed-series validation: 19 FE regressions, 32 reader tests, six C++ translation-unit compilations, Checkstyle, and BE architecture checks pass.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [x] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [x] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

Documentation for user-visible execution is included in #74663; this series targets `main` only.

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

## Related Issue

Fixes #74665

This PR implements one layer of the linked feature. The complete implementation requires all five PRs in the series listed above.

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

File: build.sh [hunk 1]
```diff
@@ -820,6 +827,9 @@ if [ ${BUILD_BE} -eq 1 ]; then
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/kudu-reader-lib ${STARROCKS_OUTPUT}/be/lib/
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/starrocks-kudu-reader.jar ${STARROCKS_OUTPUT}/be/lib/jni-packages
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/starrocks-kudu-reader.jar ${STARROCKS_OUTPUT}/be/lib/kudu-reader-lib
+        cp -r -p ${STARROCKS_HOME}/java-extensions/lance-reader/target/lance-reader-lib ${STARROCKS_OUTPUT}/be/lib/
+        cp -r -p ${STARROCKS_HOME}/java-extensions/lance-reader/target/starrocks-lance-reader.jar ${STARROCKS_OUTPUT}/be/lib/jni-packages
+        cp -r -p ${STARROCKS_HOME}/java-extensions/lance-reader/target/starrocks-lance-reader.jar ${STARROCKS_OUTPUT}/be/lib/lance-reader-lib
```
[the remark this answers is not in the document]
Comment (beinan) [reply build.sh]: Valid point. Will add existence guard for the lance-reader directory copy.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
Fixed P2 (Honor --without-connector-lance) in 835c3ac. Now checks `WITH_CONNECTOR_LANCE=ON` before copying lance-reader artifacts.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: build.sh [hunk 2]
```diff
@@ -820,6 +827,11 @@ if [ ${BUILD_BE} -eq 1 ]; then
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/kudu-reader-lib ${STARROCKS_OUTPUT}/be/lib/
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/starrocks-kudu-reader.jar ${STARROCKS_OUTPUT}/be/lib/jni-packages
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/starrocks-kudu-reader.jar ${STARROCKS_OUTPUT}/be/lib/kudu-reader-lib
+        if [ -d ${STARROCKS_HOME}/java-extensions/lance-reader/target ]; then
```
[the remark this answers is not in the document]
Comment (beinan) [reply build.sh]: Fixed in 835c3ac. Now checks WITH_CONNECTOR_LANCE=ON before copying lance-reader artifacts.

**Review by kevincai — DISMISSED** [review]

**Comment by CelerData-Reviewer** [comment]
@codex review

File: build.sh [hunk 3]
```diff
@@ -820,6 +827,11 @@ if [ ${BUILD_BE} -eq 1 ]; then
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/kudu-reader-lib ${STARROCKS_OUTPUT}/be/lib/
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/starrocks-kudu-reader.jar ${STARROCKS_OUTPUT}/be/lib/jni-packages
         cp -r -p ${STARROCKS_HOME}/java-extensions/kudu-reader/target/starrocks-kudu-reader.jar ${STARROCKS_OUTPUT}/be/lib/kudu-reader-lib
+        if [ "${WITH_CONNECTOR_LANCE}" = "ON" ] && [ -d ${STARROCKS_HOME}/java-extensions/lance-reader/target ]; then
```
[the remark this answers is not in the document]
Comment (beinan) [reply build.sh:842]: Expected behavior for now — the lance-reader module is added in PR #74662. Once that merges, the target directory will exist and the copy will execute. The existence guard prevents build failure when PRs merge in any order. See #74665 for the full dependency chain.

**Comment by beinan** [comment]
@kevincai Conflicts have been resolved and the Lance PR stack has been rebased onto the latest main. This build-system PR follows #74663. Could you please take another look when available? CI is rerunning on the updated commits.

**Comment by CelerData-Reviewer** [comment]
@codex review

File: java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceColumnValue.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        if (vector instanceof DateDayVector) {
+            int daysSinceEpoch = ((DateDayVector) vector).get(rowIndex);
+            return LocalDate.ofEpochDay(daysSinceEpoch);
+        }
+        return LocalDate.ofEpochDay(0);
+    }
+
+    @Override
+    public LocalDateTime getDateTime(ColumnType.TypeValue type) {
+        if (vector instanceof TimeStampMicroVector) {
```
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceColumnValue.java]: Fixed in 3113c53f4b, inherited by this branch. Timestamp values are decoded according to the Arrow unit for seconds, milliseconds, microseconds, and nanoseconds instead of falling back to the Unix epoch.

File: java-extensions/pom.xml [hunk 5]
```diff
@@ -21,6 +21,7 @@
         <module>odps-reader</module>
         <module>iceberg-metadata-reader</module>
         <module>kudu-reader</module>
+        <module>lance-reader</module>
```
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/pom.xml]: Fixed in fb5ff7a83a. lance-reader is now in a default-active Maven profile that can be disabled with -DskipLanceReader, and build.sh passes that property to both clean and package when WITH_CONNECTOR_LANCE=OFF. Both default and opt-out reactors were verified.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
@kevincai Review feedback has been addressed in the updated stack. fb5ff7a83a makes the Lance Maven module honor --without-connector-lance, while preserving the default build. All review threads are resolved and the PR is mergeable. Please take another look when available.

**Comment by beinan** [comment]
Rebased onto the latest main (conflict-free) and fixed both CI failures.

**FE UT (`StatisticsCalculatorTest.testValidConnectorOutputRowCount`)** — this was a genuine test bug, not flakiness. `Statistics.Builder#setOutputRowCount` clamps its argument into `[1, MAXIMUM_ROW_COUNT]` (added upstream in #67777), so `setOutputRowCount(POSITIVE_INFINITY)` becomes `MAXIMUM_ROW_COUNT` and `setOutputRowCount(-1)` becomes `1` — both finite and non-negative. The assertions expecting `false` for those two cases were therefore unreachable through the builder. The test now asserts the clamping contract instead, and keeps the case the Lance fallback actually depends on: a connector metadata that does not override `getTableStatistics` returns a builder-default `Statistics` whose row count is `NaN`, which `hasValidOutputRowCount` correctly rejects.

**BE incremental coverage (6/84)** — added `LanceTableDescriptorTest` in `be/test/runtime/descriptors_test.cpp`, covering thrift parsing, both `__isset` fallbacks that yield an empty dataset URI, and `DescriptorTbl::create` dispatch for `TTableType::LANCE_TABLE`.

The Java-Extensions coverage fix (missing `@{argLine}`, which silently disabled JaCoCo for `lance-reader`) is inherited from #74662.

**Comment by beinan** [comment]
@kevincai — apologies: your earlier review got dismissed automatically when I force-pushed the rebase onto latest main. That was not intentional, and I could not re-request you via the API (no permission), hence this ping.

All your build-system feedback has been addressed and is still in the branch:
- `--without-connector-lance` is now honored at **packaging** time, not just compile time (the copy block checks `WITH_CONNECTOR_LANCE=ON`).
- The `lance-reader/target` copy is guarded by an existence check, so the build does not break regardless of the order the stack merges in.
- The Maven reactor honors the opt-out via a default-active profile disabled by `-DskipLanceReader`, which `build.sh` passes for both `clean` and `package`.

This is the **top of the stack**, so it merges last: #74660 → #74661 → #74662 → #74663 → **#74664**. No action needed until the ones below land — flagging it mainly so the dismissed review does not get lost.

CI note: a `BE Architecture` failure on the previous run was mine — a BE unit test I added pulled in `common/config.h`, which is outside the module-boundary allowlist. The include was unnecessary (the `chunk_size` argument is unused by `DescriptorTbl::create`) and has been removed; CI is re-running.

**Comment by kevincai** [comment]
@beinan This is a large volume of code change, can be easily conflicted while other PRs merged. Wondering if this can be as small as possible with some stub code change and merged quickly and then the real code change can be made and reviewed in its own pace.
